### DISCUSSION TITLE
Is it possible to make it so refit exceptions aren't exposed to callers?

### DISCUSSION THREAD

**Discussion by ChristopherHaws** [reactiveui/refit#1470] [Q&A] [answered]
Hello,

I am using Refit in a shared library and was wondering if it is possible to wrap `ApiException` and `ValidationApiException` so that users of the library don't need to know that Refit is being used under the hoods. As far as I can see, the only way to handle exceptions right now is via `ExceptionFactory`, but even with this being implemented, Refit exceptions can still get thrown. Is the only way really to wrap the refit interface in another class and handle each exception manually?

I did find this PR which seems to be about adding the ability to customize this functionality, but it is pretty old now:
https://github.com/reactiveui/refit/pull/927

BTW, I would be ok if the solution required my custom exception to inherit `ApiException` or `ValidationApiException`, I am just hoping to make it so that our services that use the rest client nuget package (which uses refit under the hoods) doesn't expose refit specific exceptions in case we need to change to another library for some reason in the future. For example, if we switch away from using restful services for server-to-server communication and move to an event bus or gRPC, it would be nice if our client library could be updated without the services using it needing to change their code.

Thanks!

**Comment by glennawatson** [comment, marked as answer]
Two layers here, and you need both for full isolation:

1. HTTP error responses (4xx/5xx): `ExceptionFactory` already lets you return your own exception type instead of `ApiException`/`ValidationApiException`:

```csharp
var settings = new RefitSettings
{
    ExceptionFactory = async response =>
    {
        if (response.IsSuccessStatusCode) return null;
        var content = response.Content is null
            ? null
            : await response.Content.ReadAsStringAsync();
        return new MyLibraryException((int)response.StatusCode, content);
    }
};
```

It returns `Task<Exception?>`; return null for success, your own exception otherwise. Refit throws exactly what you return, so callers never see `ApiException`.

2. Everything else still leaks: serialization failures, `HttpRequestException` (DNS/connection/TLS), timeouts (`TaskCanceledException`/`OperationCanceledException`). `ExceptionFactory` does not cover these. `DeserializationExceptionFactory` covers deserialization specifically, but network/cancellation exceptions are raised by `HttpClient` itself, below Refit.

So for a clean public surface where Refit is a swappable implementation detail, the robust answer is still to wrap the Refit interface in your own service class and translate exceptions there. `ExceptionFactory` reduces the wrapping for HTTP errors, but it cannot intercept transport-level exceptions. The wrapper is one small adapter per API and fully decouples consumers from Refit, which is what you want if you might swap to gRPC / an event bus later.

You do not need your custom exception to inherit `ApiException` - it can be any `Exception`.
